Adds the generation of public roads linking towns together

VacuumBreather opened the pull request:
## Motivation / Problem

The map looks empty and unrealistic without a road network. it also makes a gamestart with a simple bus route expensive.

## Description

This change adds the generation of a public road network to the map creation process.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

stormcone commented on src/lang/english.txt:
```diff
@@ -1901,6 +1901,12 @@ STR_CONFIG_SETTING_TREES_AROUND_SNOWLINE_HELPTEXT               :Adjust placemen
 STR_CONFIG_SETTING_TREES_AROUND_SNOWLINE_RANGE_HELPTEXT         :Approximate range of arctic trees around snow line
 
+STR_CONFIG_SETTING_BUILD_PUBLIC_ROADS                           :Build public roads connecting towns: {STRING2}
+STR_CONFIG_SETTING_BUILD_PUBLIC_ROADS_HELPTEXT                  :Generates public roads which connect the times. Takes a bit of time on bigger maps. 'Build and avoid' generates roads which avoid curves and result in very grid-like connections.
```
```suggestion
STR_CONFIG_SETTING_BUILD_PUBLIC_ROADS_HELPTEXT                  :Generates public roads which connect the towns. Takes a bit of time on bigger maps. 'Build and avoid' generates roads which avoid curves and result in very grid-like connections.
```

reldred commented:
I'm feeding this through the usual collection of ridiculous maps I like to play on in order to break this and I have a couple of comments I'd like to make as I rattle my way through:

- Make sure there's a 'give up' point if the pathfinder can't find a way to get from A to B. I've been frozen now on step 352/641 for about 15mins now on a 2k x 4k heightmap with maxheight 255. I know it's still chugging away trying to find a path as the hamster wheel (CPU util) is still turning. I don't think this is unreasonable either, far flung remote towns in Alaska and Russia for instance are almost unreachable without a plane or a specialized offroad vehicle.
- I'm not sure how you're prioritizing linking towns yet but one AI/gamescript I saw did something that I liked and focused first on linking cities to their closest towns, then linking cities to other cities prioritizing routes that pass through towns, then kinda rolled the dice and linked some towns together separately afterwards. This resulted in a nice coverage that kinda 'made sense' but didn't hog the CPU for hours trying to join absolutely everything to everything else.
- See if there's a way to multithread it. I know multithreading _anything at all_ in OpenTTD is pretty new but this public road gen and then River gen (especially after I was done horsing around with it) are the two biggest bottlenecks in map gen at the moment - with the ridiculous maps I play at least.

Superb work otherwise! I've been hanging out for something like this for a while!

reldred commented:
Well, I managed to get it to assert, it's taken a couple of hours of horsing with it though.

![image](https://user-images.githubusercontent.com/19522569/121795474-78bc7a00-cc50-11eb-867f-0324fcf451e4.png)

VacuumBreather commented:
That's probably the same stupid bug I had with tunnels. The game is trying to build two bridges crossing eachother. Since it can't detect that during path finding I cannot prevent that. Hmm.. limiting it to only one bridge also it kind of problematic though... and at that point it's too late to call the path invalid... it's already trying to build it. Hmmm. I'll have to think about how to solve that. Maybe keep track of the bridges already planned and checking if a newly planned one intersects with it...

VacuumBreather commented:
> * Make sure there's a 'give up' point if the pathfinder can't find a way to get from A to B. I've been frozen now on step 352/641 for about 15mins now on a 2k x 4k heightmap with maxheight 255. I know it's still chugging away trying to find a path as the hamster wheel (CPU util) is still turning. I don't think this is unreasonable either, far flung remote towns in Alaska and Russia for instance are almost unreachable without a plane or a specialized offroad vehicle.

There actually are limits, but I guess they're not enough. Can you send me some parameters of crazy maps that drive the algorihm crazy?

Also linking towns to cities first is actually a good idea.

By the way the algorithm isn't trying to connect everything to everything. It just tries to make sure the entire map is connected in some way. It basically builds networks of connected towns. Everytime a town cannot be connected to an existing network, it creats its own new network. And it tries to connect every new unconnected town to one of those networks.

I'll go over the algorithm again though. I know it's not fast enough yet whenever there's a town on an island that cannot be connected. And since I don't allow any terraforming to happen during this, there's plenty ways for towns to be isolated.

> Superb work otherwise! I've been hanging out for something like this for a while!

Thanks. I think a very basic road network already in place just makes the maps more realistic and interesting.

VacuumBreather commented:
As for multithreading.. I'm not sure how to get that to work. All the pathfinding depends on the current state of the world. if two roads are searched for at the same time, they search against an invalid world. Once one path is built, the other one will have a shorter path to the nearest road. Parallelizing that is not really possible.

VacuumBreather commented:
I totally forgot that I originally replaced the containers used in the pathfinder... the original implementation spends 75% of the time just in a stupid hash lookup (that's why you use the standard libary folks... containers written by professionals... sigh) Pushing that fix soon. It's massively faster now.

reldred commented:
Nice! I'll give it another test!

reldred commented:
Cpt.Klutz' Cliffy Coastline 1 is my main torture test heightmap for just about any feature, it's what I used to beat on the river gen system.

Usually run it at 1k x 2k with 127 max height or 2k x 4k with 255 max height.

Also I can absolutely confirm that the new changes dramatically improve generation time. I've got some more thoughts/comments to make specifically around bridges and tunnels as they're being used currently but alas... I'm pretty drunk right now. I'll just close off by saying; I dig it, big time.

VacuumBreather commented:
Good to hear :) Let's see how we can further improve it.

reldred commented:
Can still reproduce the assert with the latest changes.

![image](https://user-images.githubusercontent.com/19522569/121811841-ca8eef80-cca4-11eb-99c6-033d042c6e4c.png)

embeddedt commented:
I've been looking for this feature in OpenTTD for a long time. Kudos. :slightly_smiling_face: 

I did find one edge case. It seems to generate bridges over land in some cases:

![image](https://user-images.githubusercontent.com/42941056/121811858-b52dad00-cc33-11eb-8229-b4a56d4124f4.png)

VacuumBreather commented:
> Can still reproduce the assert with the latest changes.
 
Yeah I haven't looked into the assert yet.

VacuumBreather commented:
Love when it creates stuff like this =)

![image](https://user-images.githubusercontent.com/61427715/121826394-bbfe0380-ccb7-11eb-86ea-1b69f0f1400d.png)

reldred commented:
Generation is looking better in some places, a lot less random long bridges in the middle of nowhere, and we're now getting some interesting formations like this:

![image](https://user-images.githubusercontent.com/19522569/121847429-b8a26080-cd27-11eb-8f17-700beda91e14.png)

But we're also getting some bizarre behavior where it's doubling over it's own routes poorly:

![image](https://user-images.githubusercontent.com/19522569/121847575-f0110d00-cd27-11eb-8f53-0296596c4a8d.png)
![image](https://user-images.githubusercontent.com/19522569/121847670-18007080-cd28-11eb-81f2-c2e0374fbaf7.png)
![image](https://user-images.githubusercontent.com/19522569/121847695-23539c00-cd28-11eb-9f16-20e9addc7b19.png)

reldred commented:
Oh yeah as a sidenote performance is now gucci, I had no problems generating maps on my i3 lenovo x220

VacuumBreather commented:
I honestly have no idea why it does that doubling over. Somehow I don't trust that A* algorithm... there's no reason it should be doing that.

reldred commented:
Nice. Your latest commit fixed it. I don't know about that assert, it only popped up twice in about a hundred map gens and I haven't been able to reproduce it since your last two commits, but I'd say this one looks just about good to go from a gameplay perspective. Some more options around the pathfinder would be nice, but this feels ready to play with right off the bat.

VacuumBreather commented:
That assert can still happen, I'll try to get a fix for it soon. But then it should be ready yes.

VacuumBreather commented:
Should be ready now. Maybe the line in FindPath() "finder.max_search_nodes = 1 << 20;" should be turned into an advanced setting. I cannot really set this much higher without a performance hit, but on 4k and bigger maps with less than 300 towns on them it is not guaranteed to link everything up if the distances are  abit too great.
Maybe a "force linkup everything" setting which sets max_search_nodes to 0 (infinity) to force the pathfinder to do an exhaustive search. That will cost performance but a player can choose to go drink a couple coffees while the map is generated if he absolutely needs those roads. For the default case, 20 should be fine.. if there's good town density it could possibly even be lowered to << 18. but with 150 towns on 2k maps I something had disconnected networks with << 18. 20 almost always guarantees a full linkup even on 8k maps unless it's very mountainous and there's few towns.

reldred commented:
Nice, I'll probably jump in and see what fine tuning I can do after it gets merged. Also not a bug report, but I am enjoying some of the construction the pathfinder is engaging in:
![image](https://user-images.githubusercontent.com/19522569/122056754-166aa180-ce29-11eb-86be-17120164447c.png)

As a side note, the path finder doesn'tr seem to be aware that roads can be built on foundations, I take it this has been skipped for the moment to keep things simple?

reldred commented:
There is one very rare glitch with the tunnels that can occasionally happen: 
![image](https://user-images.githubusercontent.com/19522569/122057511-dce66600-ce29-11eb-93ce-34da3f7f0623.png)

VacuumBreather commented:
Yeah... the problem with that one is that it's almost impossible to prevent. Before the tunnel is actually built thos tiles are not marked as such.. so I'd have to walk back the entire current_path to see if there was a tunnel planned there... I don't think that'll be very performant... there's no real other way to mark that as a tunnel tile. I#m already back-walking to prevent the tunnel-bridge assert to make sure I don't intersect any. But I cannot do that for every road tile...

VacuumBreather commented:
> As a side note, the path finder doesn'tr seem to be aware that roads can be built on foundations, I take it this has been skipped for the moment to keep things simple?

yes. Because once that foundation is built there, it prevents other things. And you face the problem yet again that you're about to change terrain but you cannot detect that during the pathfinding process yet. So you'd have to back-walk to detect those foundations. If they were already built that'd be easy. But during the pathfinding it doesn't exist yet. THe road is only built after a path has been found. That complicates matters.

reldred commented:
> 
> 
> > As a side note, the path finder doesn'tr seem to be aware that roads can be built on foundations, I take it this has been skipped for the moment to keep things simple?
> 
> yes. Because once that foundation is built there, it prevents other things. And you face the problem yet again that you're about to change terrain but you cannot detect that during the pathfinding process yet. So you'd have to back-walk to detect those foundations. If they were already built that'd be easy. But during the pathfinding it doesn't exist yet. THe road is only built after a path has been found. That complicates matters.

Gotcha. I had a feeling there was some complications under the hood.

> 
> 
> Yeah... the problem with that one is that it's almost impossible to prevent. Before the tunnel is actually built thos tiles are not marked as such.. so I'd have to walk back the entire current_path to see if there was a tunnel planned there... I don't think that'll be very performant... there's no real other way to mark that as a tunnel tile. I#m already back-walking to prevent the tunnel-bridge assert to make sure I don't intersect any. But I cannot do that for every road tile...

It's rare enough a glitch it can probably just be ignored.

Well, otherwise, I can't fault it from a players perspective. I've fed it small maps, big maps, flat maps, height maps and terragenesis map, and the straight up stupid maps I tend to play on and it's held up. I can't speak for the code side, JGR has to wipe my backside there still, but from a players perspective it feels good to go.

James103 commented:
Have you tried this with super skinny maps, such as 256x1024, 128x4096, or even 64x65536?

reldred commented:
> 
> 
> Have you tried this with super skinny maps, such as 256x1024, 128x4096, or even 64x65536?

I've done plenty of 4:1 aspect maps (I personally play a lot of them) and it works nicely, I'm gonna try the most extreme case now and see if I can break it.

VacuumBreather commented:
Actually the tunnel glitch should be gone. I didn't check which direction I arrive at tunnel entrances from. Should be good now.

VacuumBreather commented:
I changed a few things in the getNeighbors  fuction and the cost function. I introduced anouther assert and think I got rid of it. But if you guys can give this another stress test I'd appreciate it.

JGRennison commented:
I've made a few changes for now, I'll carry on looking at this tomorrow.

I've added support for trivial foundations (3 corners raised and 2 opposite corners raised) because building/using these doesn't prevent building any other road pieces on the tile.
Road and bridges can be built on coast tiles. This enables bridges over the sea, I've given bridges over the sea a higher length limit.
I've increased the cost of tunnels and bridges such that they're less likely to be used unnecessarily.
Loading of JokerPP savegames is fixed (it previously loaded without anything obviously wrong, but settings following the public roads setting got bytes from other settings).

VacuumBreather commented:
> I've added support for trivial foundations (3 corners raised and 2 opposite corners raised) because building/using these doesn't prevent building any other road pieces on the tile.

Ah gotcha, forgot about those yes :)

> Road and bridges can be built on coast tiles. This enables bridges over the sea, I've given bridges over the sea a higher length limit.

Stupid me. I totally forgot that not including cost tiles rules out those bridges.

> Loading of JokerPP savegames is fixed (it previously loaded without anything obviously wrong, but settings following the public roads setting got bytes from other settings).

Yeah that needed fixing, I was still not 100% sure how all that feature check code of yours works so I left that to you and worked on the performance instead. If you kind more ways to improve that great :)

VacuumBreather commented:
Hmm, from testing I think you should lover the cost for tunnels again. it builds roads over unnecessarily high peaks now and builds basically no tunnels at all anymore. That doesn't look like a realistically built road network anymore.
I think the cost should only be added for bridges. Tunnels were fine before I think.

VacuumBreather commented:
In the mean-time, forced the pathfinder to build serpentine roads:

![image](https://user-images.githubusercontent.com/61427715/122313860-4abc8b80-cf17-11eb-82bf-8d40904277e8.png)

VacuumBreather commented:
Yeah that tunnel cost change needs to go. We now get stuff like this:

![image](https://user-images.githubusercontent.com/61427715/122314722-0500c280-cf19-11eb-82da-8364afe127ea.png)

reldred commented:
Is there any way to add a cost to excessive amounts of tiles that climb upwards in one go?

KeldorKatarn commented:
Yes I can but it's already expensive. The problem is the new tunnel punishment. If I now make slopes in turn even more costly then the pathfinder needs to check too many nodes to find an alternative and the performance goes into the toilet again. I just tried that.

VacuumBreather commented:
> Is there any way to add a cost to excessive amounts of tiles that climb upwards in one go?

Yes I can but it's already expensive. The problem is the new tunnel punishment. If I now make slopes in turn even more costly then the pathfinder needs to check too many nodes to find an alternative and the performance goes into the toilet again. I just tried that.

reldred commented:
> 
> 
> > Is there any way to add a cost to excessive amounts of tiles that climb upwards in one go?
> 
> Yes I can but it's already expensive. The problem is the new tunnel punishment. If I now make slopes in turn even more costly then the pathfinder needs to check too many nodes to find an alternative and the performance goes into the toilet again. I just tried that.

Got it. I just did a fresh checkout and had a play with the generation and overall it's looking pretty good apart from the rarity of tunnels now. Performance on an anemic little Surface Go tablet is pretty good as well. I don't think 'perfect' is ever going to be achievable but already map generation with your recent patches is now almost a generational jump between JGRPP and OpenTTD now.

One last final thing I'd like to see based on my testing; can a final pass be done to look at the 'ends' of each network leg, look at the distance between them and the nearest town on a seperate network or network leg, and if they're close enough to attempt to join them? While it's a very minor complaint, I've seen a few networks that have turned into giant horseshoes where the two 'ends' of each network have ended up being very close to each other. The Migrations Gamescript has been helping with joining some of these legs together as well as joining industries (it also tends to do all of its construction with the dirt roads of the NRT pack I'm using which is quite pretty).

VacuumBreather commented:
I know that is an issue but right now I'm traversing the towns form the map center to the outside, trying to progressively link up towns. That distance ordering from the initial start is a performance optimization. If I link up towns out of order just because it looks pretty that messes with this algorithm.
What it COULD do is make a post-processing path and detect towns that are close together but not connected by a short road connection. but for that I'd have to make a completely new pathfinder check with different functions that only allow road connections. I mean I guess for that I could use the vehicle route finder.

But remember on bigger maps there's a LOT of such town pairs... That's a lot of checking and I really don't want to add even more pathfinding checks on exactly those maps where generating these roads already took the longest to begin with.

So I really don't want to 'prettify' this, for performance reasons. THe fact that some roads are sub-optimal and there's better routes between some towns was also a very conscious decision behind this feature. After all, there should still be something to do for the player to optimize routes. That's what the gameplay is for after all. If I just make every town connected by ideal routes already, there's no game left to play with roads. That's also why I do not connect industries with this.

VacuumBreather commented:
> Got it. I just did a fresh checkout and had a play with the generation and overall it's looking pretty good apart from the rarity of tunnels now. Performance on an anemic little Surface Go tablet is pretty good as well. I don't think 'perfect' is ever going to be achievable but already map generation with your recent patches is now almost a generational jump between JGRPP and OpenTTD now.

Did you also check out the latest serpentine addition? Does that look good in your optinion?

reldred commented:
No, that's all fair. I might jump in after this has gotten merged and have a tinker with it and look at some other out of scope features that I'm interested in.

Serpentine's look gorgeous, overall the results that it's producing at the moment - rarity of tunnels aside - are beautiful.

reldred commented:
Getting the occasional loops, this is on a 1k x 2k map with towns set to very low, map height 255, heightmap is Cliffy Coastline 1 by Cpt.K (probably my favorite heightmap author for punishing OTTD terrain gen):

![image](https://user-images.githubusercontent.com/19522569/122338906-81c68780-cf7f-11eb-902e-fa4c73c96880.png)

All in all though, I can put with this, it's not too ugly all things considered.

VacuumBreather commented:
I noticed those loops also. I'm not at all sure what is causing them... had to be the serpentine code... but I'm not sure why that would cause it... it's definitely a bug that needs fixing still.

VacuumBreather commented:
That loop bug got introduced by JGR's changes. I'm not sure how yet.

VacuumBreather commented:
I reverted that change while keeping the fix for the redundant distance check. I don't yet know WHY it broke the pathfinder, but it did. The performance sometimes dropped incredibly low on mountainous maps, the pathfinder drew loops and sometimes even build roads completely into nowhere.

Just to clarify. The pathfinder SHOULD build interesting things like this:

![image](https://user-images.githubusercontent.com/61427715/122352152-241a4700-cf4f-11eb-8b63-b843b9feea2a.png)

VacuumBreather commented:
@reldred can you stress test this again to confirm those broken paths are gone?

auge8472 commented:
Speaking about costs. How does your patch take speed limits as well as building and maintenance costs of road types into account?

I ask because in my (not released) road set I provide multiple non-town-roadtypes for overland traffic with different speed limits and costs and availabilities over time. There's a slow and cheap accomodation road, a faster country road and, beginning with 1930, a even faster and expensive motorway/highway. I *for myself* would prefer to use the country road (IMHO the best compromise between speed and costs) but other people will and the code could have a different opinion.

VacuumBreather commented:
The roads are built by the deity owner,so they don't cost anybody anything. And it uses the road type the first town in the list uses. Any upgrades should be made by the player.

reldred commented:
> @reldred can you stress test this again to confirm those broken paths are gone? (Btw small three tile loops in a T-junction are not a bug. That's three road tiles that already exist, it just fills in the circle when passing through there in another direction.)

I'm on it. Doing a fresh compile of this and improved_trees now.

reldred commented:
Getting quite a bit of what almost looks like route duplication here causing lots of loops/doubling, road generation otherwise is quite pretty:

![image](https://user-images.githubusercontent.com/19522569/122382623-404ad200-cfa9-11eb-81c4-27a27c7e1fff.png)

![image](https://user-images.githubusercontent.com/19522569/122383012-99b30100-cfa9-11eb-8445-1b297dbf98c4.png)

![image](https://user-images.githubusercontent.com/19522569/122383118-b0595800-cfa9-11eb-8457-7480ced9a183.png)

VacuumBreather commented:
> Getting quite a bit of what almost looks like route duplication here causing lots of loops/doubling, road generation otherwise is quite pretty:

That's AFTER I reverted that one commit? Dammit. I don't see that much duplication. I didn't see any... only with that tunnel/bridge limit... are you sure you built it with my reverted commit?

reldred commented:
> 
> 
> > Getting quite a bit of what almost looks like route duplication here causing lots of loops/doubling, road generation otherwise is quite pretty:
> 
> That's AFTER I reverted that one commit? Dammit. I don't see that much duplication. I didn't see any... only with that tunnel/bridge limit... are you sure you built it with my reverted commit?

Yes sorry I deleted my post, I noticed something screwy happening with Git in VS2019. Doing fresh recompiles and will re-test.

VacuumBreather commented:
@reldred If you want to connect two more towns later in your own build, you just need to call the FindPath function with a AyStar finder and the two towns. if it finds a path it will immediately build the road in the FoundEndNode callback. All you need to do is identify the two towns you want to connect. FindPath should do the rest. It might not always build a new road, but if the existing one is really THAT much longer, it should build a new one.

JGRennison commented:
The issue where tunnels were built coming through the middle of an existing town road should be fixed now.

On larger maps in particular, the road layout being a completely acyclic spanning tree is quite noticeable, it breaks the illusion a bit. (I would expect that this will be the first thing that users ask for/comment on).
Deliberately adding a few cycles, perhaps where the ratio of road distance to direct distance is highest, might be an idea.

I've been doing a bit of experimenting with pathfinder costs. My previous commit with bridges/tunnels was too punitive. A modest cost increase where bridges/tunnels cost 2x or even 3x per unit length compared with ordinary road improves results at least in my eyes. At the moment rather long tunnels are built even for very mild obstacles. I will do some more twiddling on this.

I did try penalising two sloped tiles directly adjacent to each other a bit more but I don't think it made much of a difference.

Nitpicking aside, I'm definitely intending to include this in the next release.

VacuumBreather commented:
Just make sure we don't break the pathfinder again and cause the stupid looping :D A* is easy to break with the h and g function

VacuumBreather commented:
I added some cycles. Please stress test. I hope it doesn't cost that much performance. it does definitely take longer. About twice as long. But I guess the result might be worth it.

reldred commented:
It's great, I love it. I'll keep beating on it and see if I can break it.

VacuumBreather commented:
Had to adjust the cost function again, otherwise no shortcuts would have been built after an existing connection was already established.

VacuumBreather commented:
One more change to adjust bridge/tunnel use and slope avoiding.

VacuumBreather commented:
![image](https://user-images.githubusercontent.com/61427715/122592946-41d9d000-d065-11eb-83b6-5e25738aa3cb.png)

![image](https://user-images.githubusercontent.com/61427715/122592978-4acaa180-d065-11eb-8fb3-a7ae7c6f417b.png)

![image](https://user-images.githubusercontent.com/61427715/122593142-87969880-d065-11eb-8ab8-e18d15dfc719.png)

![image](https://user-images.githubusercontent.com/61427715/122593209-a1d07680-d065-11eb-8474-356ece95bdef.png)

![image](https://user-images.githubusercontent.com/61427715/122593413-ed832000-d065-11eb-9989-89ca5dadc535.png)

![image](https://user-images.githubusercontent.com/61427715/122593670-4ce13000-d066-11eb-8543-be12246a8917.png)

JGRennison commented:
Many thanks for this.

I did a rebase to remove the duplicate commits. Looks to be from merging an earlier rebase.
